amethyst_renderer/src/cam.rs

```diff
@@ -53,25 +53,128 @@ impl From<Projection> for Matrix4<f32> {
 }
 
 /// Camera struct.
-///
-/// TODO: Add more convenience methods, refine API.
 #[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
 pub struct Camera {
+    /// Initial forward vector of the camera.
+    pub base_forward: Vector3<f32>,
+    /// Calculated center point of the camera.
+    center: Vector3<f32>,
+    /// Trigger re-calculations if values were changed.
+    pub dirty: bool,
     /// Location of the camera in three-dimensional space.
     pub eye: Point3<f32>,
     /// Graphical projection of the camera.
     pub proj: Matrix4<f32>,
-    /// Forward vector of the camera.
-    pub forward: Vector3<f32>,
-    /// Right vector of the camera.
-    pub right: Vector3<f32>,
+    /// Rotation of the camera in three-dimensional space.
+    pub rotation: Quaternion<f32>,
     /// Upward elevation vector of the camera.
     pub up: Vector3<f32>,
 }
 
 impl Camera {
+    /// Create a new camera from the given values.
+    pub fn new(
+        projection: Matrix4<f32>,
+        position: Point3<f32>,
+        look_at: Point3<f32>,
+        up: Vector3<f32>
+    ) -> Self {
+        if position == look_at {
+            panic!("Position and LookAt cannot have the same coordinates!");
```

Comment:
I think Amethyst should only panic in very special situations, which most likely don't even occur. Just think about the fact that this data might come from a config file or a script, which shouldn't cause a panic.

Reply:
@minecrawler In instances where the structure creation can fail it's typically considered appropriate to have new return a `Result` or `Option`